File: rust_impl/src/analysis/entropy.rs
pub fn calculate_entropy(data: &[u8]) -> f64 {
    if data.is_empty() {
        return 0.0;
    }

    let mut frequencies = [0usize; 256];
    for &byte in data {
        frequencies[byte as usize] += 1;
    }

    let len = data.len() as f64;
    let mut entropy = 0.0;

    for count in frequencies {
        if count == 0 {
            continue;
        }
        let probability = count as f64 / len;

        entropy -= probability * probability.log2();
    }
    entropy
}
pub fn max_entropy_for_length(len: usize) -> f64 {
    if len <= 1 {
        return 0.0;
    }
    let distinct_cap = len.min(256) as f64;
    distinct_cap.log2()
}

File: Python_Impl/src/heuristics/entropy.py
import math
from pathlib import Path

def calculate_entropy(data: bytes) -> float:
    if not data:
        return 0.
    byte_counts = [0] * 256
    for byte in data:
        byte_counts[byte] += 1
    entropy = 0.0
    total_bytes = len(data)

    for count in byte_counts:
        if count > 0:
            probability = count / total_bytes
            entropy -= probability * math.log2(probability)

    return round(entropy, 4)

def check_file_entropy(file_path: str, threshold: float = 7.2) -> dict:

    try:

        path = Path(file_path)
        if not path.is_file():
            return {"error": "File not found", "is_suspicious": False}

        binary_data = path.read_bytes()
        entropy_score = calculate_entropy(binary_data)

        is_suspicious = entropy_score >= threshold

        return {
            "file_name": path.name,
            "entropy": entropy_score,
            "is_suspicious": is_suspicious,
            "description": "High entropy detected (likely packed or encrypted)" if is_suspicious else "Normal entropy signature"
        }
    except Exception as e:
        return {"error": str(e), "is_suspicious": False}
